server/src/service.rs

```diff
@@ -1,30 +1,76 @@
 // org.freedesktop.Secret.Service
 
-use std::collections::HashMap;
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
 
-use oo7::dbus::{
-    api::{Properties, SecretInner},
-    Algorithm, ServiceError,
+use oo7::{
+    dbus::{
+        api::{Properties, SecretInner},
+        Algorithm, ServiceError,
+    },
+    Key,
 };
+use tokio::sync::RwLock;
 use zbus::{
     proxy::ProxyDefault,
     zvariant::{ObjectPath, OwnedObjectPath, OwnedValue, Value},
 };
 
+use crate::{service_manager::ServiceManager, session::Session};
+
 pub type Result<T> = std::result::Result<T, ServiceError>;
 
 #[derive(Debug, Default)]
-pub struct Service {}
+pub struct Service {
+    manager: Arc<Mutex<ServiceManager>>,
+    session_index: RwLock<i32>,
+}
 
 #[zbus::interface(name = "org.freedesktop.Secret.Service")]
 impl Service {
     #[zbus(out_args("output", "result"))]
     pub async fn open_session(
         &self,
-        _algorithm: Algorithm,
-        _input: Value<'_>,
+        algorithm: Algorithm,
+        input: Value<'_>,
+        #[zbus(object_server)] object_server: &zbus::ObjectServer,
     ) -> Result<(OwnedValue, OwnedObjectPath)> {
-        todo!()
+        let (public_key, aes_key) = match algorithm {
+            Algorithm::Plain => (None, None),
+            Algorithm::Encrypted => {
+                let client_public_key = Key::from(input);
+                let private_key = Key::generate_private_key();
+                (
+                    Some(Key::generate_public_key(&private_key)),
+                    Some(Key::generate_aes_key(&private_key, &client_public_key)),
+                )
+            }
+        };
+
+        let mut sessions = *self.session_index.read().await;
+        sessions += 1;
+        *self.session_index.write().await = sessions;
```

Comment:
Write the new value after usage. Also I am curious why you use RwLock from Tokio and Mutex from std.

Reply:
>  Also I am curious why you use RwLock from Tokio and Mutex from std.

I read a bunch of reddit posts explaining what's better and decided to go with their findings. I thought to myself you'll catch this if this is wrong and I'm glad that you did :)

What should I go with? std or tokio?

Reply:
I would have to read about it to give you a concrete answer, but just use whatever we use in the client so it is consistent

Reply:
Now both Mutex, RwLock are from tokio. I think this is ok for now. I'll continue to read on this and will update it if necessary.